tests/csapi/room_upgrade_test.go

```diff
@@ -0,0 +1,309 @@
+package csapi_tests
+
+import (
+	"testing"
+	"time"
+
+	"github.com/tidwall/gjson"
+
+	"github.com/matrix-org/complement/internal/b"
+	"github.com/matrix-org/complement/internal/client"
+	"github.com/matrix-org/complement/internal/match"
+	"github.com/matrix-org/complement/internal/must"
+)
+
+// sytest: /upgrade creates a new room
+func TestUpgradeRoom(t *testing.T) {
+	deployment := Deploy(t, b.BlueprintAlice)
+	defer deployment.Destroy(t)
+
+	alice := deployment.Client(t, "hs1", "@alice:hs1")
+
+	roomID := alice.CreateRoom(t, map[string]interface{}{
+		"preset": "public_chat",
+	})
+
+	_, tombstoneSinceToken := alice.MustSync(t, client.SyncReq{TimeoutMillis: "0"})
+	_, sinceToken := alice.MustSync(t, client.SyncReq{TimeoutMillis: "0"})
+
+	resp := alice.MustDoFunc(
+		t,
+		"POST",
+		[]string{"_matrix", "client", "v3", "rooms", roomID, "upgrade"},
+		client.WithJSONBody(t, map[string]string{
+			"new_version": "6",
+		}),
+	)
+
+	upgradeJson := gjson.ParseBytes(must.ParseJSON(t, resp.Body))
+	must.MatchGJSON(
+		t,
+		upgradeJson,
+		match.JSONKeyPresent("replacement_room"),
+		match.JSONKeyTypeEqual("replacement_room", gjson.String),
+	)
+	replacementRoom := upgradeJson.Get("replacement_room").Str
+
+	if roomID == replacementRoom {
+		t.Fatalf("Rooms are the same (%s)", roomID)
+	}
+
+	// Check for and fetch the tombstone event ID
+	tombstoneEventId := ""
+	alice.MustSyncUntil(t, client.SyncReq{Since: tombstoneSinceToken}, client.SyncTimelineHas(roomID, func(ev gjson.Result) bool {
+		if ev.Get("type").Str == "m.room.tombstone" && ev.Get("state_key").Exists() {
+			repRoomRes := ev.Get("content.replacement_room")
+
+			if repRoomRes.Exists() {
+				tombstoneEventId = ev.Get("event_id").Str
+
+				if repRoomRes.Str != replacementRoom {
+					t.Errorf("tombstone event did not match replacement room, got %s, want %s", repRoomRes.Str, replacementRoom)
+				}
+
+				return true
+			}
+		}
+
+		return false
+	}))
+
+	// Convenience method to check for state event
+	stateExists := func(evType string) func(gjson.Result) bool {
+		return func(ev gjson.Result) bool {
+			return ev.Get("type").Str == evType && ev.Get("state_key").Exists()
+		}
+	}
+
+	// Note: we cannot check for the following state here, as it's only prescribed as an "implementation detail" by the spec:
+	// - m.room.guest_access
+	// - m.room.history_visibility
+	// - m.room.join_rules
+	alice.MustSyncUntil(t, client.SyncReq{Since: sinceToken}, client.SyncJoinedTo(alice.UserID, replacementRoom),
+		client.SyncTimelineHas(replacementRoom, stateExists("m.room.create")),
+		client.SyncTimelineHas(replacementRoom, stateExists("m.room.member")),
+		client.SyncTimelineHas(replacementRoom, stateExists("m.room.power_levels")),
```

Comment:
The old sytest used to assert that the create event was the first thing to appear in the timeline. Don't mind dropping that from this test, but just wanted to highlight that it's missing.

Reply:
There's not exactly an easy way to assert that, and honestly I'd use a complement-library-defined function for it instead of assembling one here.

I'll drop a todo 👍

Reply:
I also note in passing that the spec is unusually precise about the expected order of events that /createRoom should create: https://spec.matrix.org/v1.5/client-server-api/#post_matrixclientv3createroom . So it is surprising that we don't have the same specificity when it comes to the initial batch of events in an upgraded room. https://spec.matrix.org/v1.5/client-server-api/#server-behaviour-16